From 885e5e7ed85a33af4512bfad5ada9a1986fa1a66 Mon Sep 17 00:00:00 2001
From: Yu Watanabe <watanabe.yu+github@gmail.com>
Date: Fri, 11 May 2018 15:43:04 +0900
Subject: [PATCH] network: make route_update() accept NULL

This also fixes a wrong argument for route_configure().

Fixes #8960.

(cherry picked from commit 26db55f39e436d8c2dae46b85e1bc068bfe1fd1d)
---
 src/network/networkd-dhcp6.c |  4 +++-
 src/network/networkd-route.c | 10 ++++------
 2 files changed, 7 insertions(+), 7 deletions(-)

diff --git a/src/network/networkd-dhcp6.c b/src/network/networkd-dhcp6.c
index 234e0a4602..f82e915399 100644
--- a/src/network/networkd-dhcp6.c
+++ b/src/network/networkd-dhcp6.c
@@ -202,11 +202,13 @@ static int dhcp6_pd_prefix_distribute(Link *dhcp6_link, Iterator *i,
                 if (r < 0)
                         return r;
 
+                route->family = AF_INET6;
+
                 while (n < n_prefixes) {
                         route_update(route, &prefix, pd_prefix_len, NULL, NULL,
                                      0, 0, RTN_UNREACHABLE);
 
-                        r = route_configure(route, link, NULL);
+                        r = route_configure(route, dhcp6_link, NULL);
                         if (r < 0) {
                                 route_free(route);
                                 return r;
diff --git a/src/network/networkd-route.c b/src/network/networkd-route.c
index 70dca5219b..8446feb316 100644
--- a/src/network/networkd-route.c
+++ b/src/network/networkd-route.c
@@ -379,14 +379,12 @@ void route_update(Route *route,
                  unsigned char type) {
 
         assert(route);
-        assert(src);
-        assert(gw);
-        assert(prefsrc);
+        assert(src || src_prefixlen == 0);
 
-        route->src = *src;
+        route->src = src ? *src : (union in_addr_union) {};
         route->src_prefixlen = src_prefixlen;
-        route->gw = *gw;
-        route->prefsrc = *prefsrc;
+        route->gw = gw ? *gw : (union in_addr_union) {};
+        route->prefsrc = prefsrc ? *prefsrc : (union in_addr_union) {};
         route->scope = scope;
         route->protocol = protocol;
         route->type = type;
-- 
2.17.0

